initial ITC integration

tpolecat opened the pull request:
First, I apologize for the size of this PR. There's no avoiding it really.

-  I added an `itc` package that contains an implementation of `Itc[F]` that is backed by remote calls to an OCS2 ITC server. Most of the code in the PR is just supporting classes and serialization code to make the model in `itc` line up with what the server expects. The model isn't a copy/paste of the OCS2 … I cleaned up a few things that make it easier to follow on our end but requires some custom logic to serialize in the right way. I don't know if this was the right choice but it's self-contained and easy to change if we want to. Note that the stuff in `itc` is private to the implementation; the OCS3/basic-case model is independent.

- I added some missing data types and syntax extensions for existing ones, to provide extra information we need for ITC (mostly serialization keys to map from OCS3 to OCS2).

- There's a sanity test that checks the three main cases (single exposure, multiple exposure, no solution).

I haven't hooked it up with the mode search mechanism yet, and I had to stub out an `ImaginaryItc` to keep the sequence stuff compiling. I will work on these things next.

Once it's all working I will push the enums and other generic stuff back into OCS3.

swalker2m commented on src/main/scala/TargetProfile.scala:
```diff
@@ -0,0 +1,19 @@
+final case class TargetProfile(
+  spatialProfile:       SpatialProfile,
+  spectralDistribution: SpectralDistribution,
+  magnitude:            Double,
```
I've never understood magnitudes and systems very well. I wonder if there is a way to represent magnitude as an integer?  I suppose it depends on the system.  I wonder if there should be three kinds of magnitude with their own brightness value instead of a `magnitude` `Double` and a `MagnitudeSystem` `enum`.

tpolecat replied:
There are actually six magnitude systems, plus an additional six with the units squared for surface brightness. We only have three in the OT but ITC supports all six and from talking with Andy I think we need to as well.

A problem here is that the units are all extremely weird and I haven't found a way to represent them. They can probably all be scaled `Long`s but I'd like to return to this as a separate task.

tpolecat replied:
We have a `MagnitudeValue` class in OCS3 but it's half-baked. I like the ADT idea.

swalker2m commented on src/main/scala/itc/ItcObservationDetails.scala:
```diff
@@ -0,0 +1,204 @@
+      final case class Imaging(
+        exposures:      Int,
+        coadds:         Option[Int],
+        exposureTime:   Double,
```
Is there a reason we need raw `Double`s for these things instead of, say, a Java `Duration` or a Scala `FiniteDuration`?

tpolecat replied:
ITC speaks in terms of `Double` but I can switch to use `FiniteDuration` here. It's just a question of how much we want to diverge from the OCS2 model down here in the glue code.

swalker2m commented on src/test/scala/itc/Test.scala:
```diff
@@ -0,0 +1,67 @@
+    )
+
+  def withItc(f: Itc[IO] => IO[Assertion]): Assertion =
+    ItcImpl.forHeroku[IO].use(f).unsafeRunSync
```
I'm uneasy about seeing unit tests that depend on external resources. We do need tests but it would be nice to have the option of skipping anything that requires a network connection and an external service to be running.

tpolecat replied:
Yeah this is temporary. I'll add that "slow test" flag.

tpolecat commented:
Ok I removed the `Double` durations and was able to simplify some things so that was a good suggestion. I also removed a `Double` arcseconds in favor of `Angle` while I was at it.

swalker2m commented:
Okay that looks great! 👍 Unfortunately you'll have to rebase and reapply the "imaginaryItc" update to the sequence gen.  Sorry about that!

tpolecat commented:
No worries.
